app.py

```diff
@@ -1461,6 +1594,68 @@ def rec(tree, n=0, depth=0):
 				(x, thistree))
 	return nodes + ''.join(leaves), estimator.tree_.max_depth, path
 
+def merge_punctuation(tokens):
```

Comment:
As discussed, it seems preferable to store the punctuation tokenization from the graphical tree in a separate data structure when punctuation is stripped out, rather than merging punctuation into a string with the adjacent lexeme just to be heuristically segmented later.

Reply:
Addressed in newly-defined `tree_process()` function. 

Example usage:

```
Here 's a sentence I wrote ... _. with lots of little marks ( called " punctuation " ) . . .
```

(Note periods are separately tokenized at end, to indicate they don't form an ellipsis token)

Dopparser parse:

<img width="1522" alt="Screenshot 2024-08-08 at 9 47 39 PM" src="https://github.com/user-attachments/assets/41c334d9-7511-4029-8bc4-14c34dc4c708">